core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala

```diff
@@ -434,18 +453,61 @@ private[group] class GroupMetadata(val groupId: String, initialState: GroupState
     }.toMap
   }
 
-  def removeExpiredOffsets(startMs: Long) : Map[TopicPartition, OffsetAndMetadata] = {
-    val expiredOffsets = offsets
-      .filter {
-        case (topicPartition, commitRecordMetadataAndOffset) =>
-          commitRecordMetadataAndOffset.offsetAndMetadata.expireTimestamp < startMs && !pendingOffsetCommits.contains(topicPartition)
-      }
-      .map {
-        case (topicPartition, commitRecordOffsetAndMetadata) =>
-          (topicPartition, commitRecordOffsetAndMetadata.offsetAndMetadata)
-      }
+  def removeExpiredOffsets(currentTimestamp: Long, offsetRetentionMs: Long) : Map[TopicPartition, OffsetAndMetadata] = {
+    val expiredOffsets: Map[TopicPartition, CommitRecordMetadataAndOffset] = protocolType match {
+      case Some(protocol) if is(Empty) =>
+        // no consumer exists in the group =>
+        // if retention period has passed since group became Empty, expire all offsets with no pending offset commit
+        offsets.filter { case (topicPartition, commitRecordMetadataAndOffset) =>
+            !pendingOffsetCommits.contains(topicPartition) && {
+              commitRecordMetadataAndOffset.offsetAndMetadata.expireTimestamp match {
+                case OffsetCommitRequest.DEFAULT_EXPIRATION_TIMESTAMP =>
+                  // current version with no per partition retention
+                  currentTimestamp - currentStateTimestamp >= offsetRetentionMs
+                case expireTimestamp =>
+                  // older versions with explicit expire_timestamp field => old expiration semantics is used
+                  currentTimestamp >= expireTimestamp
+              }
+            }
+        }.toMap
+
+      case Some(protocol) =>
+        // there are active consumers in the group =>
+        // expire offsets of partitions the group is no longer subscribed to, assuming retention period
+        // has passed since their last commit
+        val subscribedTopics = allMemberMetadata.flatMap { memberMetadata =>
+          memberMetadata.supportedProtocols.flatMap { supportedProtocol =>
+            ConsumerProtocol.deserializeSubscription(ByteBuffer.wrap(supportedProtocol._2)).topics().asScala.toSet
```

Comment:
This only works if the protocol type is "consumer." I'd suggest leaving for another patch. We need to think through a more general approach for adding behavior to the coordinator which depends on the protocol type.

Reply:
I would suggest that we leave subscription checking out of this patch and implement it in a follow-up. For now, maybe we can just have expiration for all offsets begin when the group becomes empty.  What I am imagining is basically a general mechanism for "installing" protocol type behavior, but it may take a few iterations to get to something that makes sense, so I'd rather not couple it with the rest of this.

Reply:
Ok, thanks for clarifying. This means this PR would become Part 1 of KIP-211 implementation, as we are leaving a piece of it out (i.e. no additional KIP would be required when we come back to implementing this piece). Correct me if I'm wrong.

Reply:
Yes, right.